---
title: "BARRY"
author: "Oisin McCaffrey"
date: "6/21/2021"
output: html_document
---

```{r}
library(VariantAnnotation)
library(stringr)
library(dplyr)
library(ggplot2)
library(EbayesThresh)
library(pinfsc50)
library(vcfR)
library(dplyr)
```

```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("VariantAnnotation")
```

```{r}
library(VariantAnnotation)
```

```{r}

```

```{r}
info(vcf)
```

```{r}

```


```{r}
VariantAnnotation::row
```

```{r}
row
```

```{r}
library(tibble)
```




```{r}
library(VariantAnnotation)
vcf <- readVcf("/Users/oisinmccaffrey/Desktop/R_Shiny_Summer/SRR.gavin_secondpass.vcf", "GRCh38")
vcf_1 <- as.data.frame(VariantAnnotation::fixed(vcf))
vcf_2 <- as.data.frame(VariantAnnotation::info(vcf))
vcf_3 <- as.data.frame(rowRanges(vcf))
# duplicate info here, will let you decide whats important
vcf_master <- cbind(vcf_3, vcf_1, vcf_2)
# stage an empty DF which we will populate in the for loop.
# If you want to extract another col, be sure to add it here
collect_ann <- data.frame(Allele=as.character(),
                          Consequence=as.character(),
                          IMPACT=as.character(),
                          Symbol=as.character(),
                          Gene=as.character())
# loop over every row in df
for(i in 1:nrow(vcf_master)){
  # grab the ann column
  ann <- vcf_master$ANN[i]
  # parse ann column
  ann <- str_split(ann, "\\|")
  # convert to DF 
  ann <- as.data.frame(t(unlist(ann)))
  # we want the first 5 cols in this case
  ann <- ann[,1:5]
  # rename to match the collect_ann df
  colnames(ann) <- c("Allele", "Consequence", "IMPACT", "Symbol", "Gene")
  # populate the collect ann df
  collect_ann <- rbind(collect_ann, ann)
}
# append to master df
vcf_master <- cbind(vcf_master, collect_ann)
# I do not know the official terms for these columns, figure that out
collect_rlv <- data.frame(Status=as.character(),
                          METIN=as.character(),
                          Prediction=as.character(),
                          CADD_Relevance=as.character())
for(i in 1:nrow(vcf_master)){
  rlv <- vcf_master$RLV[i]
  rlv <- str_split(rlv, "\\|")
  rlv <- as.data.frame(t(unlist(rlv)))
  rlv <- rlv[,c(11,13,15,17)]
  colnames(rlv) <- c("Status", "METIN", "Prediction", "CADD_Relevance")
  collect_rlv <- rbind(collect_rlv, rlv)
}
vcf_master <- cbind(vcf_master, collect_rlv)

vcf_master$Allele <- gsub('^c\\(|\\)$', '', vcf_master$Allele)
vcf_master$Allele <- gsub("[^A-Za-z0-9]", "", vcf_master$Allele)
vcf_master <- vcf_master %>% tibble::rownames_to_column(var = "ID_")

is.na(vcf_master$ID_) <- startsWith(vcf_master$ID_, "1") | startsWith(vcf_master$ID_, "2") | startsWith(vcf_master$ID_, "3") | startsWith(vcf_master$ID_, "4") | startsWith(vcf_master$ID_, "5") | startsWith(vcf_master$ID_, "6") | startsWith(vcf_master$ID_, "7") | startsWith(vcf_master$ID_, "8") | startsWith(vcf_master$ID_, "9") | startsWith(vcf_master$ID_, "10") | startsWith(vcf_master$ID_, "11") | startsWith(vcf_master$ID_, "12") | startsWith(vcf_master$ID_, "13") | startsWith(vcf_master$ID_, "14") | startsWith(vcf_master$ID_, "15") | startsWith(vcf_master$ID_, "16") | startsWith(vcf_master$ID_, "17") | startsWith(vcf_master$ID_, "18") | startsWith(vcf_master$ID_, "19") | startsWith(vcf_master$ID_, "20") | startsWith(vcf_master$ID_, "21") | startsWith(vcf_master$ID_, "22") | startsWith(vcf_master$ID_, "X")

```

```{r}
library(dplyr)
```

```{r}
vcf_master
```






```{r}
vcf_master_validID
```




```{r}
head(rowRanges(vcf), 3)
```


```{r}
DS <-geno(vcf)$DS 
dim(DS)
```




```{r}
genes <- vcf_master %>% select(c(Symbol, Gene, seqnames, start, REF, Allele, Consequence, IMPACT))
genes <- dplyr::rename(genes, Chr = seqnames, From = REF, To = Allele, HGNC = Symbol)
```

```{r}
vcf_master$EXAC_AF <- as.numeric(vcf_master$EXAC_AF)
```


```{r}
vcf_master$EXAC_AF <- log(vcf_master$EXAC_AF)
```

vcf_master[vcf_master$Consequence=='missense_variant'] <- 'Missense Variant'

```{r}
cadd_id_plot <- ggplot(data = vcf_master, 
                 aes(x= as.factor(EXAC_AF), y = as.factor(CADD_SCALED), fill = Consequence)) + 
                geom_point(size = 2) +
                ggtitle("CADD Score vs. SNP Accession") +
                scale_y_discrete(limits = c(15, 50,
                                          breaks = c(15,  20,  30, 40, 50),
                                          labels = c("15", "20", "30", "40","50"),
                                          expand=c(0,0),
                                          name = "CADD SCORE") +
                scale_x_discrete("Variant ID") +
                theme_minimal() +
                theme(
                    axis.line.y = element_blank(),
                    axis.ticks.y = element_blank(),
                    axis.line.x = element_blank(),
                    axis.ticks.x = element_blank(),
                    axis.text.x = element_text(),
                    axis.title.x = element_blank(),
                    panel.grid.major.x =element_line(),
                    panel.grid.minor.x = element_blank(),
                    panel.grid.major.y = element_blank(),
                    panel.background=element_blank(),
                    panel.ontop = TRUE, 
                    legend.position= c(0.75, 0.92), legend.direction="horizontal",
                    legend.text = element_text(size = 6), 
                    legend.key.size = unit(0.6, "lines"),
                    legend.title = element_text(size =10)))
            
            cadd_id_plot <- cadd_id_plot + ylab("CADD Score") + 
              theme(axis.title.y = element_text(angle = 0)) +
              xlab("rs Accession ID") + 
              theme(axis.title.x = element_text(angle = 0))
            
            ggplotly(cadd_id_plot)
```






```{r}
vcf_master$Allele
```
gsub("e", "", group)

```{r}
vcf_master$Consequence <- gsub("_", " ", vcf_master$Consequence)
vcf_master$Consequence <- gsub("&", " and ", vcf_master$Consequence)
vcf_master$Consequence
```



```{r}
vcf_master$QUAL[1]
```


```{r}
vcf_master$CADD_Relevance[1]
```


```{r}
df <-as.data.frame(vcf_master$QUAL)
df <- rename(df, QUAL = `vcf_master$QUAL`)
df <- transform(df, QUAL = as.numeric(QUAL))
df
```




```{r}
qual_hist <- ggplot(vcf_master, aes(QUAL)) 

qual_hist + geom_histogram(bins = 30, color = "black", fill = "#40E0D0") +
  geom_vline(aes(xintercept = mean(QUAL)), 
             linetype = "dashed", size = 0.6) +
  geom_text(aes(x=650, label="mean line", y=53), colour="black", angle=0, text=element_text(size=6)) +
  ggtitle("Quality (QUAL)") +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5)) +
  ylab("Frequency") +
  theme(axis.title.y = element_text(angle = 0, vjust = 0.5)) 
```


```{r}
head(vcf)
```


```{r}
qual_hist <- ggplot(data=vcf_master, aes(QUAL)) 

qual_hist + geom_histogram(bins = 30, color = "black", fill = "#40E0D0") +
  geom_vline(aes(xintercept = mean(QUAL)), 
             linetype = "dashed", size = 0.6) +
  scale_x_continuous(limits = c(0, 2000), 
                     expand = c(0, 0),
                      breaks = seq(0, 2000, by = 250),
                     name = "MQ") +
  ggtitle("Quality (QUAL)") +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5)) +
  ylab("Frequency") +
  theme(axis.title.y = element_text(angle = 0, vjust = 0.5)) 
  
  
```


```{r}
df_DP<- as.data.frame(vcf_master$DP)
df_DP <- rename(df_DP, DP = `vcf_master$DP`)
df_DP <- transform(df_DP, DP = as.numeric(DP))
```


```{r}
dp_hist <- ggplot(data=df_DP, aes(DP)) 

dp_hist + geom_histogram(bins = 16, color = "black", fill = "#5CD85A") +
  ggtitle("Read Depth (DP)") +
  geom_vline(aes(xintercept = mean(DP)), 
             linetype = "dashed", size = 0.6) +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5, face = "bold")) +
  ylab("Frequency") +
  theme(axis.title.y = element_text(angle = 0, vjust = 0.5)) 
```



```{r}
df_MQ<- as.data.frame(vcf_master$MQ)
df_MQ <- rename(df_MQ, MQ = `vcf_master$MQ`)
df_MQ <- transform(df_MQ, MQ = as.numeric(MQ))
df_MQ
```



```{r}
MQ_hist <- ggplot(data=df_MQ, aes(MQ)) 

MQ_hist + geom_histogram(bins = 16, color = "black", fill = "#0059b3") +
  ggtitle("Mapping Quality (MQ)") +
  theme_minimal() +
  scale_x_continuous(limits = c(55, 65), 
                     expand = c(0, 0),
                      breaks = seq(55, 65, by = 1),
                     name = "MQ") +
  theme(plot.title = element_text(hjust = 0.5, face = "bold")) +
  ylab("Frequency") +
  theme(axis.title.y = element_text(angle = 0, vjust = 0.5)) 
```
```{r}
vcf_master
```

```{r}
vcf_master$EXAC_AF[1]
```



```{r}
library(cowplot)
```


```{r}
qual_hist_plot <- ggplot(df, aes(QUAL)) +
        geom_histogram(bins = 30, color = "black", fill = "#00FFFF") +
        geom_vline(aes(xintercept = mean(QUAL)), 
                   linetype = "dashed", size = 0.6) +
        scale_x_continuous(limits = c(0, 2000), 
                           expand = c(0, 0),
                           breaks = seq(0, 2000, by = 250),
                           name = "QUAL") +
        ggtitle("Quality (QUAL)") +
        theme_minimal() +
        theme(plot.title = element_text(hjust = 0.5, face = "bold")) +
        ylab("Frequency") +
        theme(axis.title.y = element_text(angle = 0, vjust = 0.5)) 

qual_hist_plot
```

    

      
```{r}
dp_hist_plot <- ggplot(data=df_DP, aes(DP)) +
   geom_histogram(bins = 16, color = "black", fill = "#5CD85A") +
        ggtitle("Read Depth (DP)") +
        theme_minimal() +
        geom_vline(aes(xintercept = mean(DP)), 
                   linetype = "dashed", size = 0.6) +
        theme(plot.title = element_text(hjust = 0.5, face = "bold")) +
        ylab("Frequency") +
        theme(axis.title.y = element_text(angle = 0, vjust = 0.5)) 
      
      
      dp_hist_plot
```

      
```{r}
MQ_hist_plot <- ggplot(data=df_MQ, aes(MQ)) +
        geom_histogram(bins = 16, color = "black", fill = "#0059b3") +
        ggtitle("Mapping Quality (MQ)") +
        theme_minimal() +
        scale_x_continuous(limits = c(55, 65), 
                           expand = c(0, 0),
                           breaks = seq(55, 65, by = 1),
                           name = "MQ") +
        theme(plot.title = element_text(hjust = 0.5, face = "bold")) +
        ylab("Frequency") +
        theme(axis.title.y = element_text(angle = 0, vjust = 0.5))
      
      MQ_hist_plot
```



#use cowplot:plot_grid for grid plot layout
g_grid <- plot_grid(g2, g1, labels = "AUTO")
g_grid


```{r}
metrics_grid <- plot_grid(dp_hist_plot, MQ_hist_plot, qual_hist_plot, QD_hist_plot, labels = "AUTO")
```

```{r}
metrics_grid
```

```{r}
df_QD<- as.data.frame(vcf_master$QD)
df_QD <- rename(df_QD, QD = `vcf_master$QD`)
df_QD <- transform(df_QD, QD = as.numeric(QD))
df_QD
```


```{r}
vcf_master$Status
```


```{r}
QD_hist_plot <- ggplot(data=df_QD, aes(QD)) +
        geom_histogram(bins = 16, color = "black", fill = "#FF69B4") +
        ggtitle("Quality by Depth (QD)") +
        theme_minimal() +
        theme(plot.title = element_text(hjust = 0.5, face = "bold")) +
        ylab("Frequency") +
        theme(axis.title.y = element_text(angle = 0, vjust = 0.5))
      
      QD_hist_plot
```
```{r}
vcf_master <- vcf_master %>% drop_na(ID_)
```


```{r}
library("SNPlocs.Hsapiens.dbSNP144.GRCh38")
snps <- SNPlocs.Hsapiens.dbSNP144.GRCh38
snpsById(snps, vcf_master$ID_, ifnotfound="drop")
```


```{r}
vcf_master$seqnames
```

```{r}
vcf_master <- vcf_master %>% drop_na(seqnames)
```


```{r}
ID_plot_dataframe <- vcf_master
ID_plot_dataframe <- as.data.frame(ID_plot_dataframe)
ID_plot_dataframe <- transform(ID_plot_dataframe, ID_ = as.character(ID_))
ID_plot_dataframe <- transform(ID_plot_dataframe, seqnames = as.numeric(seqnames))
ID_plot_dataframe
```


```{r}
ID_location_plot <- ggplot(vcf_master, 
                             aes(x=ID_, y = seqnames, fill=Status)) + 
  coord_flip() +
        geom_point() +
  theme(
         axis.line.y = element_blank(),
         axis.ticks.y = element_blank(),
         axis.line.x = element_blank(),
         axis.ticks.x = element_blank(),
         axis.text.x = element_text(),
         axis.title.x = element_blank(),
        panel.background = element_blank(),
        panel.grid = element_blank(),
        plot.margin = unit(c(0.5, 0.5, 2, 0.5), "cm"),
        plot.title = element_text(size=8, face="bold")) +
  scale_x_discrete("Accession ID") +
  scale_y_discrete("Genomic Location") +
  theme(
    legend.text = element_text(size = 8),
    legend.position = c(0.55, -0.2), # move to the bottom
    legend.title = element_blank(),
    legend.key.size = unit(0.9, "line"),
    legend.spacing.x = unit(0.2, 'cm'),
    legend.background = element_rect(
      fill = "white",
      size = 0.5,
      colour = "white"
    )
    ) 

ID_location_plot <- ID_location_plot + theme(axis.title.x = element_text(angle = 0)) 
ID_location_plot <- ID_location_plot + theme(axis.text.y = element_blank())
ID_location_plot <- ID_location_plot + theme(axis.ticks.y = element_blank())

ggplotly(ID_location_plot)
```

```{r}
library(VariantAnnotation)
```

```{r}

```


```{r}
summarizeVariants(vcf)
```

```{r}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
```

```{r}
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
```

```{r}
vcf1 <- readVcf("/Users/oisinmccaffrey/Desktop/R_Shiny_Summer/SRR.gavin_secondpass.vcf", "hg38")
## Rename seqlevels to match TxDb; confirm the match.
seqlevels(vcf1) <- paste0("chr", seqlevels(vcf1))
intersect(seqlevels(vcf1), seqlevels(txdb))
```


```{r}
## ----------------------------------------
## Counting with locateVariants()
## ----------------------------------------
## TxDb as the 'query'
coding1 <- summarizeVariants(txdb, vcf1, CodingVariants()) 
colSums(assays(coding1)$counts)
## GRangesList as the 'query'
cdsbytx <- cdsBy(txdb, "tx")
coding2 <- summarizeVariants(cdsbytx, vcf1, CodingVariants())
stopifnot(identical(assays(coding1)$counts, assays(coding2)$counts))
## Promoter region variants summarized by transcript
tx <- transcripts(txdb)
txlst <- splitAsList(tx, seq_len(length(tx)))
promoter <- summarizeVariants(txlst, vcf,
PromoterVariants(upstream=100, downstream=10))
colSums(assays(promoter)$counts)
```



```{r}

vcf_master <- dplyr::rename(vcf_master, Chr = seqnames)
```

```{r}
vcf_master_plots <- vcf_master %>% drop_na(ID_)
vcf_master_plots <- vcf_master %>% drop_na(CADD_SCALED)
vcf_master_plots <- vcf_master %>% drop_na(EXAC_AF)
vcf_master_plots <- vcf_master %>% drop_na(Chr)



# transform CADD_SCALED column to be numeric data
vcf_master_plots <- transform(vcf_master_plots, CADD_SCALED = as.numeric(CADD_SCALED))
vcf_master_plots <- transform(vcf_master_plots, EXAC_AF = as.numeric(EXAC_AF))

vcf_master_plots <- transform(vcf_master_plots, ID_ = as.character(ID_))
```

```{r}
vcf_master_plots
```




```{r}
vcf_master_plots_CADD <- vcf_master_plots %>% drop_na(CADD_SCALED)
vcf_master_plots_CADD
```

ifelse(x > median(x),"up","down"))



CLICK





```{r}
vcf_master_plots$Other <- ifelse(is.na(vcf_master_plots$CADD_SCALED), "Other", "Has CADD Score")
```

```{r}
vcf_master_plots$Other
```

```{r}
vcf_master_plots_CADD_Others <- vcf_master_plots[vcf_master_plots$Other != "Has CADD Score", ]
vcf_master_plots_CADD_Others
```

```{r}
library(stringr)
vcf_master_plots_CADD_Others$CADD_SCALED_SCORE <- str_extract(vcf_master_plots_CADD_Others$CADD_Relevance, '(?i) (?<=score of\\D)\\d+')

vcf_master_plots_CADD_Others
```


```{r}
vcf_master_plots_CADD_Others
```

```{r}
new_dataset <- vcf_master_plots %>% dplyr::select(c(ID_, Chr, Status, CADD_SCALED, Consequence, EXAC_AF))
```

```{r}
library(dplyr)
library(stringr)
library(tidyr)
```

```{r}
vcf_master_plots_CADD_Others <- vcf_master_plots_CADD_Others %>% dplyr::select(c(ID_, Chr, Status, CADD_SCALED_SCORE, Consequence, EXAC_AF))
vcf_master_plots_CADD_Others
```

```{r}
vcf_master_plots_CADD_Others <- dplyr::rename(vcf_master_plots_CADD_Others,
                                              ID_Other = ID_,
                                              Chr_Other = Chr,
                                              Status_Other = Status,
                                              CADD_SCALED_OTHER = CADD_SCALED_SCORE,
                                              Consequence_Other = Consequence,
                                              EXAC_AF_Other = EXAC_AF)
```

```{r}
vcf_master_plots_CADD_Others
```

```{r}
new_dataset <- dplyr::bind_rows(vcf_master_plots_CADD_Others, vcf_master_plots)
```


```{r}
new_dataset <- transform(new_dataset, CADD_SCALED_OTHER = as.numeric(CADD_SCALED_OTHER))
```

```{r}
new_dataset <- new_dataset %>% unite("CADD_ALL", CADD_SCALED_OTHER, CADD_SCALED)
new_dataset <- new_dataset %>% unite("ID_ALL", ID_, ID_Other)
new_dataset <- new_dataset %>% unite("EXAC_AF_ALL", EXAC_AF_Other, EXAC_AF)
new_dataset <- new_dataset %>% unite("Consequence_ALL", Consequence_Other, Consequence)
new_dataset <- new_dataset %>% unite("Status_ALL", Status, Status_Other)
new_dataset <- new_dataset %>% unite("Chr_ALL", Chr_Other, Chr)
```


```{r}
new_dataset$CADD_ALL <- sub("_NA", " ", new_dataset$CADD_ALL)
new_dataset$CADD_ALL <- sub("NA_", " ", new_dataset$CADD_ALL)
new_dataset$ID_ALL <- sub("_NA", " ", new_dataset$ID_ALL)
new_dataset$ID_ALL <- sub("NA_", " ", new_dataset$ID_ALL)
new_dataset$EXAC_AF_ALL <- sub("_NA", " ", new_dataset$EXAC_AF_ALL)
new_dataset$EXAC_AF_ALL <- sub("NA_", " ", new_dataset$EXAC_AF_ALL)
new_dataset$Consequence_ALL <- sub("_NA", " ", new_dataset$Consequence_ALL)
new_dataset$Consequence_ALL <- sub("NA_", " ", new_dataset$Consequence_ALL)
new_dataset$Status_ALL <- sub("_NA", " ", new_dataset$Status_ALL)
new_dataset$Status_ALL <- sub("NA_", " ", new_dataset$Status_ALL)
new_dataset$Chr_ALL <- sub("_NA", " ", new_dataset$Chr_ALL)
new_dataset$Chr_ALL <- sub("NA_", " ", new_dataset$Chr_ALL)
```

```{r}
new_dataset$Status_ALL <- gsub("SRR:", " ", new_dataset$Status_ALL)
new_dataset$Status_ALL <- gsub("_", " ", new_dataset$Status_ALL)
```

```{r}
new_dataset <- transform(new_dataset, CADD_ALL = as.numeric(CADD_ALL))
```

```{r}
new_dataset <- new_dataset %>% drop_na(CADD_ALL)
```

```{r}
new_dataset
```
vcf_master_plots <- transform(vcf_master_plots, CADD_SCALED = as.numeric(CADD_SCALED))


```{r}
new_dataset1 <- transform(new_dataset, Chr_ALL = as.character(Chr_ALL))
new_dataset1
```


```{r}
chrOrder<-c(paste(1:22,sep=""),"X")
chrOrder
```



```{r}
new_dataset1$Chr_ALL <- sort(new_dataset1$Chr_ALL, levels=chrOrder)
```

df %>%
    arrange(factor(Reg, levels = LETTERS[c(3, 1, 2)]), desc(Res), desc(Pop))


```{r}
new_dataset1 %>% arrange(match(Chr_ALL, c("1", "2","3","4","5","6","7","8","9","10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "X")))
```
levels(mydt$value) <- c(...)

```{r}
levels(new_dataset1$Chr_ALL) <- c("1", "2","3","4","5","6","7","8","9","10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "X")
```

```{r}
new_dataset1
```

